// @antv/l7-map/lib/handler/box_zoom.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _dom = _interopRequireDefault(require("../utils/dom"));

var _event = require("./events/event");

// @ts-ignore

/**
 * The `BoxZoomHandler` allows the user to zoom the map to fit within a bounding box.
 * The bounding box is defined by clicking and holding `shift` while dragging the cursor.
 */
var BoxZoomHandler = /*#__PURE__*/function () {
  /**
   * @private
   */
  function BoxZoomHandler(map, options) {
    (0, _classCallCheck2.default)(this, BoxZoomHandler);
    this.map = map;
    this.el = map.getCanvasContainer();
    this.container = map.getContainer();
    this.clickTolerance = options.clickTolerance || 1;
  }
  /**
   * Returns a Boolean indicating whether the "box zoom" interaction is enabled.
   *
   * @returns {boolean} `true` if the "box zoom" interaction is enabled.
   */


  (0, _createClass2.default)(BoxZoomHandler, [{
    key: "isEnabled",
    value: function isEnabled() {
      return !!this.enabled;
    }
    /**
     * Returns a Boolean indicating whether the "box zoom" interaction is active, i.e. currently being used.
     *
     * @returns {boolean} `true` if the "box zoom" interaction is active.
     */
